packages/server/src/arpa_reporter/services/validate-upload.js

```diff
@@ -342,6 +351,121 @@ async function validateRules ({ upload, records, rules, trns }) {
 
   return errors
 }
+// Subrecipients can use either the uei, or the tin, or both as their identifier.
+// This helper takes those 2 nullable fields and converts it to a reliable format
+// so we can index and search by them.
+function subrecipientIdString(uei, tin) {
+  if (!uei && !tin) {
+    return '';
+  }
+  return JSON.stringify({ uei, tin })
+}
+
+async function validateReferences({ records }) {
+  const errors = []
+  // These 3 types need to search-able by their unique id so we can quickly verify they exist
+  const projects = {}
+  const subrecipients = {}
+  const awardsGT50k = {}
+
+  const awards = []
+  const expendituresGT50k = []
+  for (const record of records) {
+    switch (record.type) {
+      case 'ec1':
+      case 'ec2':
+      case 'ec3':
+      case 'ec4':
+      case 'ec5':
+      case 'ec7':
+        const projectID = record.content.Project_Identification_Number__c
+        if (projectID in projects) {
+          errors.push(betaValidationWarning(
+            `Project ids must be unique, but another row used the id ${projectID}`))
+        }
+        projects[projectID] = record.content;
+        break
+      case 'subrecipient':
+        const subRecipId = subrecipientIdString(
+          record.content.Unique_Entity_Identifier__c,
+          record.content.EIN__c)
+        if (subRecipId && subRecipId in subrecipients) {
+          errors.push(betaValidationWarning(
+            `Subrecipient ids must be unique, but another row used the id ${subRecipId}`))
+        }
+        subrecipients[subRecipId] = record.content
+        break
+      case 'awards50k':
+        const awardNumber = record.content.Award_No__c
```

Comment:
Is it useful to have some validation here for when the `awardNumber` is null/undefined?

Reply:
Great question!
That validation already exists earlier in the file. This field is marked as "required" in the input workbook, so if it is missing they are already getting a separate message from that other validation .I originally DID add another validation here until I saw it was redundant.